### PR TITLE
go_1_17: init at 1.17.1

### PR DESCRIPTION
**Pull request by zowoq**
###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by kalbasit** [comment]
Except from https://tip.golang.org/doc/go1.17


```
vendor contents

If the main module specifies go 1.17 or higher, go mod vendor now annotates vendor/modules.txt with the go version indicated by each vendored module in its own go.mod file. The annotated version is used when building the module's packages from vendored source code. 
If the main module specifies go 1.17 or higher, go mod vendor now omits go.mod and go.sum files for vendored dependencies, which can otherwise interfere with the ability of the go command to identify the correct module root when invoked within the vendor tree. 
```

This may change tge hashes in vendorSha256 in existing packages that are using buildGoModule.

**Comment by zowoq** [comment]
Yeah, this release isn't going to be easy ...

> Go 1.17 requires macOS 10.13 High Sierra or later

`aarch64-darwin` is `11.0.0` but `x86_64-darwin` is `10.12`.

> If the main module specifies go 1.17 or higher

Updating to 1.17 and pinning `darwin` or even just `x86_64-darwin` to 1.16 seems like it's going to end up causing problems.

**Comment by endocrimes** [comment]
> this may change the hashes of vendorSha256 in existing packages that are using buildGoModule.

this feels like a pretty good reason to not force bump everything immediately - but instead give maintainers a chance to upgrade their modules as upstreams jump to 1.17

File: pkgs/development/compilers/go/1.17.nix
Comment (endocrimes) [inline pkgs/development/compilers/go/1.17.nix:161]: Can we document why the non-test patches exist? Going history diving gets increasingly complicated as stuff moves across versions (especially deprecated ones).

I'm especially a little worried about things like the no vendor checks being enabled by default, and the SSL cert file change, because it means binaries built on nixos are not the same as those built on non-nix systems.

File: pkgs/development/compilers/go/1.17.nix
Comment (SuperSandro2000) [inline pkgs/development/compilers/go/1.17.nix:72]:
```suggestion
  postPatch = ''
```

File: pkgs/development/compilers/go/1.17.nix
Comment (SuperSandro2000) [inline pkgs/development/compilers/go/1.17.nix:73]: This should be more precise.

File: pkgs/development/compilers/go/1.17.nix
Comment (SuperSandro2000) [inline pkgs/development/compilers/go/1.17.nix:272]:
```suggestion
    homepage = "https://golang.org/";
```

**Comment by zowoq** [comment]
I'm keeping this in sync with the `go_*.nix` files so I'm not addressing nits in this PR.

File: pkgs/development/compilers/go/1.17.nix
Comment (endocrimes) [reply pkgs/development/compilers/go/1.17.nix:161]: @zowoq this wasn't a nit, it's an attempt to try and document things for people going forwards and make it easier for folks to maintain go in nix long term.

If you're not going to address it here, doing so in a separate PR to the older go versions could be useful.

**Comment by endocrimes** [comment]
I’d still like to see https://github.com/NixOS/nixpkgs/pull/127519#discussion_r656348346 at least partially resolved before this lands or soon after - it’s becoming difficult to track their history.

this explicitly is not a nitpick, but something important to longer term maintainability.

File: pkgs/development/compilers/go/1.17.nix
Comment (zowoq) [reply pkgs/development/compilers/go/1.17.nix:161]: Fell free to submit a PR and I'll rebase and copy it here.

Comment (endocrimes) [reply pkgs/development/compilers/go/1.17.nix:161]: I'm mostly assuming you have at least some of that context already - because a lot of what's important here is the why and not the what.

(Especially for things like ignoring vendor checks)

Comment (Xe) [reply pkgs/development/compilers/go/1.17.nix:161]: no vendor checks is apparently from here: https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/compilers/go/go_no_vendor_checks.patch

**Review by saschagrunert — APPROVED** [review]
LGTM

**Review by c00w — APPROVED** [review]

**Comment by c00w** [comment]
Any chance we can get this landed? It's been 2 weeks since go1.17 was released publicly and it would be great to use this without having to maintain a fork :).

**Comment by c00w** [comment]
re: updating the hashes - I still have my code (and it's somewhere on either github or git.sr.ht), which will autoregen all the hashes for go packages, so I can probably get most packages that are compatible bulk migrated once this lands.

**Comment by zowoq** [comment]
We can merge this go make `go_1_17` available but with of the go module and platform changes in this version I don't want to allow use of `buildGo117{Module,Package}` in nixpkgs for now.

@NixOS/golang  Are we okay with that and the comment I've added in `all-packages.nix`?

**Comment by terinjokes** [comment]
@zowoq Is this a ticket to track work needed to begin considering allow `buildGo117{Module,Package}`?

**Comment by Artturin** [comment]
> re: updating the hashes - I still have my code (and it's somewhere on either github or git.sr.ht), which will autoregen all the hashes for go packages, so I can probably get most packages that are compatible bulk migrated once this lands.

hey, could you send the script?

**Pull request merged**
